### ISSUE TITLE
rspec-puppet-init fails with puppet 4.0

### ISSUE THREAD

**Issue by hesco** [rodjek/rspec-puppet#290]
Before pinning puppet to v3.7.5, I was getting:

```
ROSELCDV000101K:adp_jenkins_poc escoh$ bundle exec rspec-puppet-init
/Users/escoh/sandbox/my_module/gems/ruby/2.0.0/gems/rspec-puppet-2.1.0/lib/rspec-puppet/setup.rb:60:in `get_module_name_from_file': uninitialized constant Puppet::Parser::Lexer (NameError)
```

Having done so, with this:
$ PUPPET_VERSION=3.7.5 bundle install 

$ bundle exec rspec-puppet-init 
now completes as expected.  

a conditional is required to provide a work-around for puppet 4.x.

**Comment by igalic** [comment]
is this also the case with the master of this repo?

**Comment by alexharv074** [comment]
Is the fix as simple as:

    --- /Users/alexharvey/cachedgems/ruby/2.0.0/gems/rspec-puppet-2.1.0/lib/rspec-puppet/setup.rb.orig  2015-05-26 12:28:16.000000000 +1000
    +++ /Users/alexharvey/cachedgems/ruby/2.0.0/gems/rspec-puppet-2.1.0/lib/rspec-puppet/setup.rb       2015-05-26 12:28:57.000000000 +1000
    @@ -56,7 +56,11 @@
         end
     
         def self.get_module_name_from_file(file)
    -      p = Puppet::Parser::Lexer.new
    +      if Puppet.version.to_f >= 4.0
    +        p = Puppet::Pops::Parser::Lexer2.new
    +      else
    +        p = Puppet::Parser::Lexer.new
    +      end
           module_name = nil
           p.string = File.read(file)
           tokens = p.fullscan

**Comment by alexharv074** [comment]
That said this function doesn't seem to be working for my code base, whether I use Puppet v3.3.1 or Puppet v4.1.1 - get_module_name_from_file with file manifests/site.pp as input is returning '{' - suggesting a parsing error.

**Comment by alexharv074** [comment]
I opened PR https://github.com/rodjek/rspec-puppet/pull/291

**Comment by alexharv074** [comment]
@hlindberg's comment from the PR:

"[The puppet v4 lexer2] is not documented and not marked as @api ... I think that is a telling sign. OTOH, the old Lexer was also not considered API. However, the entire method is flawed since it is only using the lexer - it is completely unknown if the read content is valid (e.g. "class class Foo define define class" lexes just fine :-) The correct way is to use the parser and get the parsed result (and look for validation errors). The definitions are found inside the parsed result. With the 3x parser evaluation was sort of intertwined with parsing, and you would have to evaluate the logic to find the definitions (it scanned for them when importing/evaluating) - and thus it was very tricky to find the definitions.

Albeit also not API, the EvaluatingParser class provides high level access to parsing, validation and evaluation. I would use it - to a) parse, b) get the Program instance (program) from the parsed result, c) get the definitions from it (program.definitions), c) iterate over those to find ResourceTypeDefinition instances, and HostClassDefinition instances, and get each definition.name d) split the name and get the first part. (Note that there are other model classes that are definitions than ResourceTypeDefinition and HostClassDefinition). See the file model_meta.rb for the schema/model of everything in the AST-model."

**Comment by alexharv074** [comment]
We agreed that I would add to my PR a FIXME with reference back to this ticket.

**Comment by logicminds** [comment]
checkout puppet-retrospect which does the same but loads more.

**Comment by jorhett** [comment]
igalic: sorry for the delay responding, Velocity and then pile of work that built up while I was gone sucked up the last two weeks.

I uninstalled all versions of the rspec-puppet gem to be certain and installed from a clone of this repo, which gave me version 2.3.0 however the error message is identical:

```
$ rake spec
/usr/bin/ruby -I/home/vagrant/.gem/ruby/gems/rspec-core-3.2.3/lib:/home/vagrant/.gem/ruby/gems/rspec-support-3.2.2/lib /home/vagrant/.gem/ruby/gems/rspec-core-3.2.3/exe/rspec --pattern spec/\{classes,defines,unit,functions,hosts,integration\}/\*\*/\*_spec.rb --color
F

Failures:

  1) puppet with defaults for all parameters should contain Class[puppet]
     Failure/Error: it { should contain_class('puppet') }
     Puppet::PreformattedError:
       Evaluation Error: Error while evaluating a Function Call, Could not find class ::puppet for client.example.com at line 1:1 on node client.example.com
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:207:in `block in evaluate_classes'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:206:in `collect'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:206:in `evaluate_classes'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/functions/include.rb:33:in `block in <top (required)>'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/functions.rb:162:in `block (2 levels) in newfunction'
 # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util/profiler.rb:51:in `profile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/functions.rb:155:in `block in newfunction'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/evaluator/runtime3_support.rb:250:in `call_function'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/evaluator/evaluator_impl.rb:836:in `call_function_with_block'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/evaluator/evaluator_impl.rb:816:in `eval_CallNamedFunctionExpression'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:46:in `block in visit_this'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:42:in `each'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:42:in `visit_this'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:69:in `visit_this_1'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/evaluator/evaluator_impl.rb:73:in `evaluate'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/evaluator/evaluator_impl.rb:640:in `eval_Program'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:46:in `block in visit_this'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:42:in `each'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:42:in `visit_this'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/visitor.rb:69:in `visit_this_1'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/evaluator/evaluator_impl.rb:73:in `evaluate'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/pops/parser/evaluating_parser.rb:57:in `evaluate'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/ast/pops_bridge.rb:100:in `evaluate'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/ast.rb:31:in `safeevaluate'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/resource/type.rb:118:in `evaluate_code'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/resource.rb:79:in `evaluate'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:409:in `evaluate_main'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:131:in `block (2 levels) in compile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util/profiler.rb:51:in `profile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:131:in `block in compile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/context.rb:64:in `override'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet.rb:223:in `override'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:121:in `compile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/parser/compiler.rb:34:in `compile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/indirector/catalog/compiler.rb:93:in `block (2 levels) in compile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util/profiler.rb:51:in `profile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/indirector/catalog/compiler.rb:91:in `block in compile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util.rb:129:in `block in benchmark'
     # /usr/share/ruby/benchmark.rb:296:in `realtime'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/util.rb:128:in `benchmark'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/indirector/catalog/compiler.rb:90:in `compile'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/indirector/catalog/compiler.rb:50:in `find'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/indirector/indirection.rb:194:in `find'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/support.rb:210:in `block in build_catalog_without_cache'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet/context.rb:64:in `override'
     # /home/vagrant/.gem/ruby/gems/puppet-4.1.0/lib/puppet.rb:223:in `override'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/support.rb:208:in `build_catalog_without_cache'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/support.rb:222:in `build_catalog'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/support.rb:27:in `load_catalogue'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/example/class_example_group.rb:7:in `catalogue'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/support.rb:7:in `block in subject'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/matchers/create_generic.rb:82:in `call'
     # /home/vagrant/.gem/ruby/gems/rspec-puppet-2.3.0/lib/rspec-puppet/matchers/create_generic.rb:82:in `matches?'
     # ./spec/classes/init_spec.rb:5:in `block (3 levels) in <top (required)>'

Finished in 10.77 seconds (files took 0.57871 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/classes/init_spec.rb:5 # puppet with defaults for all parameters should contain Class[puppet]

/usr/bin/ruby -I/home/vagrant/.gem/ruby/gems/rspec-core-3.2.3/lib:/home/vagrant/.gem/ruby/gems/rspec-support-3.2.2/lib /home/vagrant/.gem/ruby/gems/rspec-core-3.2.3/exe/rspec --pattern spec/\{classes,defines,unit,functions,hosts,integration\}/\*\*/\*_spec.rb --color failed
```

**Comment by DavidS** [comment]
This is now fixed in 2.2.0.
